addon/-private/system/relationships/state/relationship.js

```diff
@@ -155,19 +157,31 @@ export default class Relationship {
     heimdall.increment(addCanonicalRecord);
     if (!this.canonicalMembers.has(record)) {
       this.canonicalMembers.add(record);
-      if (this.inverseKey) {
-        record._relationships.get(this.inverseKey).addCanonicalRecord(this.record);
-      } else {
-        if (!record._implicitRelationships[this.inverseKeyForImplicit]) {
-          record._implicitRelationships[this.inverseKeyForImplicit] = new Relationship(this.store, record, this.key,  { options: {} });
-        }
-        record._implicitRelationships[this.inverseKeyForImplicit].addCanonicalRecord(this.record);
-      }
+      this.setupInverseRelationship(record, false);
     }
     this.flushCanonicalLater();
     this.setHasData(true);
   }
 
+  setupInverseRelationship(internalModel, initial) {
+    if (this.inverseKey) {
+      let relationships = internalModel._relationships;
+      let relationshipExisted = relationships.has(this.inverseKey);
+      let relationship = relationships.get(this.inverseKey);
+      if (relationshipExisted || this.isPolymorphic) {
```

Comment:
Can you leave comments explaining 
1. The laziness
2. why we are not lazy in the polymorphic case